arrow.go

```diff
@@ -93,26 +93,120 @@ func NewArrowFromConn(driverConn driver.Conn) (*Arrow, error) {
 	return &Arrow{conn: conn}, nil
 }
 
-var _ array.RecordReader = (*arrowStreamReader)(nil)
-
-// arrowStreamReader implements array.RecordReader for streaming DuckDB results.
-type arrowStreamReader struct {
-	ctx      context.Context
-	res      *arrowmapping.Arrow
-	schema   *arrow.Schema
-	rowCount uint64
-
-	mu         sync.Mutex // protects readCount and currentRec
-	refCount   int64
-	readCount  uint64
-	currentRec arrow.Record
-	closed     bool // tracks if the reader has been closed/released
-	err        error
+// QueryContext prepares statements, executes them, returns Apache Arrow array.RecordReader as a result of the last
+// executed statement. Arguments are bound to the last statement.
+func (a *Arrow) QueryContext(ctx context.Context, query string, args ...any) (array.RecordReader, error) {
+	if a.conn.closed {
+		return nil, errClosedCon
+	}
+
+	rows, err := a.conn.QueryContext(ctx, query, a.anyArgsToNamedArgs(args))
+	if err != nil {
+		return nil, err
+	}
+
+	return recordReaderFromRows(ctx, rows)
 }
 
-// Retain increases the reference count by 1.
-// Retain may be called simultaneously from multiple goroutines.
-func (r *arrowStreamReader) Retain() {
+// RegisterView registers an Arrow record reader as a view with the given name in DuckDB.
+// The returned release function must be called to release the memory once the view is no longer needed.
+func (a *Arrow) RegisterView(reader array.RecordReader, name string) (release func(), err error) {
+	if a.conn.closed {
+		return nil, errClosedCon
+	}
+
+	stream := C.calloc(1, C.sizeof_struct_ArrowArrayStream)
+	release = func() {
+		cdata.ReleaseCArrowArrayStream((*cdata.CArrowArrayStream)(stream))
+		C.free(stream)
+	}
+	cdata.ExportRecordReader(reader, (*cdata.CArrowArrayStream)(stream))
+
+	arrowStream := arrowmapping.ArrowStream{
+		Ptr: unsafe.Pointer(stream),
+	}
+	if arrowmapping.ArrowScan(a.conn.conn, name, arrowStream) == mapping.StateError {
+		release()
+		return nil, errors.New("duckdb_arrow_scan")
```

Comment:
Can we turn this into a bit more elaborate error? I.e., that something went wrong in the DuckDB-side? Also, can we trigger this in a test? If not, should we tell people to file a bug report if they run into it?